src/doc/trpl/error-handling.md

```diff
@@ -1,308 +1,2147 @@
+corresponding value constructor. Indeed, it is *typically* this easy.
+
+We can finally rewrite `file_double`:
+
+```rust
+fn file_double<P: AsRef<Path>>(file_path: P) -> Result<i32, CliError> {
+    let mut file = try!(File::open(file_path));
+    let mut contents = String::new();
+    try!(file.read_to_string(&mut contents));
+    let n: i32 = try!(contents.trim().parse());
+    Ok(2 * n)
+}
+```
+
+The only thing we did here was remove the calls to `map_err`. They are no
+longer needed because the `try!` macro invokes `From::from` on the error value.
+This works because we've provided `From` impls for all the error types that
+could appear.
+
+If we modified our `file_double` function to perform some other operation, say,
+convert a string to a float, then we'd need to add a new variant to our error
+type:
+
+```rust
+enum CliError {
+    Io(io::Error),
+    ParseInt(num::ParseIntError),
+    ParseFloat(num::ParseFloatError),
+}
+```
+
+And add a new `From` impl:
+
+```rust
+impl From<num::ParseFloatError> for CliError {
+    fn from(err: num::ParseFloatError) -> CliError {
+        CliError::Parse(err)
     }
+}
+```
+
+And that's it!
+
+
+### Advice for library writers
+
+Idioms for Rust libraries are still forming, but if your library needs to
+report custom errors, then you should probably define your own error type.
+It's up to you whether or not to expose its representation
+(like [`ErrorKind`](http://doc.rust-lang.org/std/io/enum.ErrorKind.html))
+or keep it hidden
+(like
+[`ParseIntError`](http://doc.rust-lang.org/std/num/struct.ParseIntError.html)).
+Regardless of how you do it, it's usually good practice to at least provide
+some information about the error beyond just its `String` representation. But
+certainly, this will vary depending on use cases.
+
+At a minimum, you should probably implement the
+[`Error`](http://doc.rust-lang.org/std/error/trait.Error.html)
+trait. This will give users of your library some minimum flexibility for
+[composing errors](#the-real-try-macro). Implementing the `Error` trait also
+means that users are guaranteed the ability to obtain a string representation
+of an error (because it requires impls for both `fmt::Debug` and
+`fmt::Display`).
+
+Beyond that, it can also be useful to provide implementations of `From` on your
+error types. This allows you (the library author) and your users to
+[compose more detailed errors](#composing-custom-error-types). For example,
+[`csv::Error`](http://burntsushi.net/rustdoc/csv/enum.Error.html)
+provides `From` impls for both `io::Error` and `byteorder::Error`.
+
+Finally, depending on your tastes, you may also want to define a
+[`Result` type alias](#the-result-type-alias-idiom), particularly if your
+library defines a single error type. This is used in the standard library
+for [`io::Result`](http://doc.rust-lang.org/std/io/type.Result.html)
+and [`fmt::Result`](http://doc.rust-lang.org/std/fmt/type.Result.html).
+
```

Comment:
It would be nice to add a small paragraph here about not using `panic`, but we don't have to do this in this PR